# CEE 201: Linear Programming with Jupyter Notebooks

## Practice Problem

In [1]:
# Import python packages numpy, matplotlib, and pulp
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline
import pulp

## Model Formulation
### Decision Variables
$x_1$ and $x_2$

### Objective Function & Constraints
#### Maximizing Objective Function
$max Z=12x_1+18x_2$

#### Subject to 

$6x_1+5x_2\leq 60$ \
$x_1+5x_2\leq 15$ \
$x_1\leq 9$ \
$x_2\leq 4$ \
$x_1 \geq 0$ \
$x_2 \geq 0$ 


### Problem Solution
#### Solver Solution
We define the class of the problem:

In [2]:
my_lp_problem = pulp.LpProblem("My LP Problem", pulp.LpMaximize)

We set the upper bounds of $x_1$ and $x_2$:

In [3]:
x1 = pulp.LpVariable('$x_1$',upBound=9, cat='Continuous')
x2 = pulp.LpVariable('$x_2$',upBound=4, cat='Continuous')

The objective function and constraints are added using the **+=** operator to the model.\
The objective function **Z** is added first and then we add all the constraints.

In [4]:
# Objective function
my_lp_problem += 12* x1 + 18* x2, "Z"

# Constraints
my_lp_problem += 6 * x1 + 5 * x2 <= 60
my_lp_problem += x1 + 3* x2 <= 15
my_lp_problem += x1>= 0
my_lp_problem += x2>= 0

### Solution
We inspect the mathematical program formulation.

In [5]:
my_lp_problem

My LP Problem:
MAXIMIZE
12*$x_1$ + 18*$x_2$ + 0
SUBJECT TO
_C1: 6 $x_1$ + 5 $x_2$ <= 60

_C2: $x_1$ + 3 $x_2$ <= 15

_C3: $x_1$ >= 0

_C4: $x_2$ >= 0

VARIABLES
-inf <= $x_1$ <= 9 Continuous
-inf <= $x_2$ <= 4 Continuous

We call the solver:

In [6]:
my_lp_problem.solve()
pulp.LpStatus[my_lp_problem.status]

'Optimal'

The optimal values of the decision variables that minimize $x_1$, $x_2$ are:

In [7]:
#Print decision variables
print(" x1* = {}".format(x1.varValue))
print(" x2* = {}".format(x2.varValue))

 x1* = 8.0769231
 x2* = 2.3076923


Print the **objective values $Z$**:

In [8]:
#Print objective function value (max Z)
print(pulp.value(my_lp_problem.objective))

138.46153859999998


The minimum objective value is approximately $138.5$.

### Sensitivity Analysis and Shadow Prices for Bindings Constraints

We call a function that reports the shadow price:

In [9]:
o = [{'name':name, 'shadow price':c.pi, 'slack': c.slack} 
     for name, c in my_lp_problem.constraints.items()]
print(pd.DataFrame(o))

  name  shadow price     slack
0  _C1      1.384615 -0.000000
1  _C2      3.692308 -0.000000
2  _C3     -0.000000 -8.076923
3  _C4     -0.000000 -2.307692


The shadow price or the change in the optimal value of the objective function for a unit of change in the RHS of binding constraint $6 x_1+ 5 x_2 <= 60$ is 1.4 and 3.7 for binding constraint $x_1 + 3x_2 <= 15$, respectively.